src/contracts/signer-identity.ts

```diff
@@ -0,0 +1,249 @@
+
+
+/**
+ * signer identity config
+ */
+export interface SignerIdentityConfig {
+  activeIdentity: string;
+  underlyingAccount: string;
+  underlyingSigner: SignerInterface;
+}
+
+/**
+ * signer identity options
+ */
+export interface SignerIdentityOptions extends LoggerOptions {
+  contractLoader: ContractLoader;
+  verifications: Verifications;
+  web3: any;
+}
+
+/**
+ * Signer, that can sign transactions for identity based transactions. If given an identity contract
+ * address as `.from`, an identity based transaction is made, if given an accountId as `.from`, a
+ * direct transaction via externally owned account is made.
+ *
+ * @class      SignerIdentity (name)
+ */
+export class SignerIdentity extends Logger implements SignerInterface {
+  public activeIdentity: string;
+  public underlyingAccount: string;
+  private coder: any = new AbiCoder();
+  private config: SignerIdentityConfig;
+  private options: SignerIdentityOptions;
+
+  public constructor(options: SignerIdentityOptions, config: SignerIdentityConfig) {
+    super(options);
+    this.options = options;
+    this.config = config;
+    this.activeIdentity = this.config.activeIdentity;
+    this.underlyingAccount = this.config.underlyingAccount;
+  }
+
+  /**
+   * Creates a smart contract.
+   *
+   * @param      {string}  contractName       contract name
+   * @param      {any[]}   functionArguments  arguments for contract creation, pass empty Array if
+   *                                          no arguments
+   * @param      {any}     options            transaction options, having at least .from and .gas
+   * @return     {Promise<any>}  new contract reference
+   */
+  public async createContract(
+    contractName: string,
+    functionArguments: any[],
+    options: any,
+  ): Promise<any> {
+    if (options.from === this.underlyingAccount) {
+      return this.config.underlyingSigner.createContract(contractName, functionArguments, options);
+    }
+    // build input for contructror call
+    const compiledContract = this.options.contractLoader.getCompiledContract(contractName);
+    if (!compiledContract || !compiledContract.bytecode) {
+      throw new Error(`cannot find contract bytecode for contract "${contractName}"`);
+    }
+    // build bytecode and arguments for constructor
+    options.input = `0x${compiledContract.bytecode}` +
+      this.encodeConstructorParams(JSON.parse(compiledContract.interface), functionArguments);
+
+    const { blockNumber, transactionHash } = await this.handleIdentityTransaction(null, null, [], options);
+    const keyHolderLibrary = this.options.contractLoader.loadContract(
+      'KeyHolderLibrary', this.activeIdentity);
+    const events = await keyHolderLibrary.getPastEvents(
+      'ContractCreated', { fromBlock: blockNumber, toBlock: blockNumber });
+    const matches = events.filter(ev => ev.transactionHash === transactionHash);
+    if (matches.length !== 1) {
+      throw new Error('contract creation failed');
```

Comment:
in which situations could this be the case?Can we make the error message more specific?

Reply:
Unfortunately not, contract creation may fail due to some inner transactions, that failed. A more detailed reason can only be found by manual analysis.